syncs/watchdog.go

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package syncs
+
+import (
+	"context"
+	"sync"
+	"time"
+)
+
+// Watch monitors mu for contention.
+// On first call, and at every tick, Watch locks and unlocks mu.
+// (Tick should be large to avoid adding contention to mu.)
+// Max is the maximum length of time Watch will wait to acquire the lock.
+// The time required to lock mu is sent on the returned channel.
+// Watch exits when ctx is done, and closes the returned channel.
+func Watch(ctx context.Context, mu sync.Locker, tick, max time.Duration) chan time.Duration {
+	// Set up the return channel.
+	c := make(chan time.Duration)
```

Comment:
What's the value in neatly closing out the return channel? As long as the goroutines are cleaned up on ctx.Done before Watch returns, it feels like the channel can be left open.

Reply:
I want callers to be able to `for d := range c` and trust that `c` will get closed when the context expires. Makes the caller's job easier, which is a good thing.

Reply:
I suppose so. I've basically given up on ranging over channels and always select in a loop, which is easy to throw a <-ctx.Done() into. Maybe that's why this API feels weird to me.

Reply:
I can understand that, but I find it a bit sad.